unittests/CAPI/Dialect/Struct.cpp

```diff
@@ -0,0 +1,353 @@
+  {
+    auto s = mlirStringRefCreateFromCString("T");
+    auto sym = mlirFlatSymbolRefAttrGet(ctx, s);
+    llvm::SmallVector<MlirAttribute> attrs(
+        {mlirFlatSymbolRefAttrGet(ctx, mlirStringRefCreateFromCString("A"))}
+    );
+    auto t = llzkStructTypeGetWithAttrs(sym, attrs.size(), attrs.data());
+    EXPECT_NE(t.ptr, (void *)NULL);
+  }
+}
+
+TEST_F(CAPITest, llzk_type_is_a_struct_type) {
+  {
+    auto s = mlirStringRefCreateFromCString("T");
+    auto sym = mlirFlatSymbolRefAttrGet(ctx, s);
+    auto t = llzkStructTypeGet(sym);
+    EXPECT_NE(t.ptr, (void *)NULL);
+    EXPECT_TRUE(llzkTypeIsAStructType(t));
+  }
+}
+
+TEST_F(CAPITest, llzk_struct_type_get_name) {
+  {
+    auto s = mlirStringRefCreateFromCString("T");
+    auto sym = mlirFlatSymbolRefAttrGet(ctx, s);
+    auto t = llzkStructTypeGet(sym);
+    EXPECT_NE(t.ptr, (void *)NULL);
+    EXPECT_TRUE(mlirAttributeEqual(sym, llzkStructTypeGetName(t)));
+  }
+}
+
+TEST_F(CAPITest, llzk_struct_type_get_params) {
+  {
+    auto s = mlirStringRefCreateFromCString("T");
+    auto sym = mlirFlatSymbolRefAttrGet(ctx, s);
+    llvm::SmallVector<MlirAttribute> attrs(
+        {mlirFlatSymbolRefAttrGet(ctx, mlirStringRefCreateFromCString("A"))}
+    );
+    auto a = mlirArrayAttrGet(ctx, attrs.size(), attrs.data());
+    auto t = llzkStructTypeGetWithArrayAttr(sym, a);
+    EXPECT_NE(t.ptr, (void *)NULL);
+    EXPECT_TRUE(mlirAttributeEqual(a, llzkStructTypeGetParams(t)));
+  }
+}
+
+struct TestOp {
+  MlirOperation op;
+
+  ~TestOp() { mlirOperationDestroy(op); }
+};
+
+class StructDefTest : public CAPITest {
+
+protected:
+  MlirOperation new_struct() const {
+    {
+      auto struct_name = mlirFlatSymbolRefAttrGet(ctx, mlirStringRefCreateFromCString("S"));
+      auto name = mlirStringRefCreateFromCString("struct.new");
+      auto location = mlirLocationUnknownGet(ctx);
+      auto result = llzkStructTypeGet(struct_name);
+      auto op_state = mlirOperationStateGet(name, location);
+      mlirOperationStateAddResults(&op_state, 1, &result);
+      return mlirOperationCreate(&op_state);
+    }
+  }
+  TestOp test_op() const {
+    {
+      auto elt_type = mlirIndexTypeGet(ctx);
+      auto name = mlirStringRefCreateFromCString("arith.constant");
+      auto attr_name = mlirIdentifierGet(ctx, mlirStringRefCreateFromCString("value"));
+      auto location = mlirLocationUnknownGet(ctx);
+      llvm::SmallVector<MlirType> results({elt_type});
+      auto attr = mlirIntegerAttrGet(elt_type, 1);
+      llvm::SmallVector<MlirNamedAttribute> attrs({mlirNamedAttributeGet(attr_name, attr)});
+      auto op_state = mlirOperationStateGet(name, location);
+      mlirOperationStateAddResults(&op_state, results.size(), results.data());
+      mlirOperationStateAddAttributes(&op_state, attrs.size(), attrs.data());
+      return {
+          .op = mlirOperationCreate(&op_state),
+      };
+    }
+  }
+};
+
+TEST_F(StructDefTest, llzk_operation_is_a_struct_def_op) {
```

Comment:
`llzkOperationIsAStructDefOp()` will always be false in all of these tests right? So these functions are just testing that the proper API functions exist? Is there a test somewhere for building a StructDefOp?

Reply:
No, there aren't any that check for StructDefOp because that op does not expose a custom build method to the C API. For the ops that do I have tests that check if the op validates after building.